File: Engine/Android/core/src/test/kotlin/despia/engine/StringsConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The localization kernel-seam conformance runner — executes
 * OpenSource/Conformance/strings/cases.json through THIS runtime's DSXStrings (the TS
 * runner and the Swift reference run the SAME file). The laws under test are the corpus
 * `_note`: the locale ladder (global.locale over the device language, `en` is the
 * source), full-tag-then-bare candidate order with the first non-empty merged table
 * winning outright, the runtime tier merging OVER the bundle tier, the (lang, version)
 * reload key (a runtime writer must bump global.strings.version to be seen), and
 * fail-open identity for every miss, absent table, invalid table JSON, and empty input.
 *
 * The seams are driven exactly as the web runner drives them: `statePath` reads the
 * case's dot-keyed state map, `loader` serves the case's bundle-table TEXT (text, so
 * invalid JSON is expressible), `deviceLang` is the case's device.
 */
class StringsConformanceTest {

    private fun corpusFile(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/strings/cases.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/strings/cases.json not found")
        }
    }

    private val savedDeviceLang = DSXStrings.deviceLang

    @AfterEach
    fun restore() {
        DSXStrings.loader = null
        DSXStrings.statePath = { null }
        DSXStrings.deviceLang = savedDeviceLang
        // force the next lookup to reload: the cache keys on (lang, version), and a
        // fresh case may resolve the same pair with different seams
        DSXStrings.localize("")
        DSXStrings.deviceLang = "__reset"
        DSXStrings.localize("x")
        DSXStrings.deviceLang = savedDeviceLang
    }

    @Suppress("UNCHECKED_CAST")
    @TestFactory
    fun corpusCases(): List<DynamicTest> {
        val doc = json(corpusFile().readText()).foundationValue as? Map<*, *>
            ?: error("strings/cases.json: not a JSON object")
        val cases = doc["cases"] as? List<Map<String, Any?>> ?: error("strings/cases.json: no cases")
        assertTrue(cases.size >= 15, "expected the full case set, found ${cases.size}")
        return cases.map { c ->
            DynamicTest.dynamicTest(c["name"] as String) {
                val state = HashMap<String, Any?>((c["state"] as? Map<String, Any?>) ?: emptyMap())
                val tables = (c["tables"] as? Map<String, Any?>) ?: emptyMap()
                // invalidate the cache across cases: a lang the runtime has never seen
                DSXStrings.statePath = { null }
                DSXStrings.deviceLang = "__cache-break-${c["name"]}"
                DSXStrings.localize("x")
                DSXStrings.loader = { lang -> tables[lang] as? String }
                DSXStrings.statePath = { path -> if (state.containsKey(path)) state[path] else null }
                DSXStrings.deviceLang = (c["device"] as? String) ?: "en"
                val steps: List<Map<String, Any?>> = (c["steps"] as? List<Map<String, Any?>>)
                    ?: listOf(mapOf("input" to c["input"], "expect" to c["expect"]))
                steps.forEachIndexed { index, step ->
                    (step["state"] as? Map<String, Any?>)?.let { state.putAll(it) }
                    assertEquals(
                        step["expect"] as String,
                        DSXStrings.localize(step["input"] as String),
                        "${c["name"]}: step ${index + 1}",
                    )
                }
            }
        }
    }
}

File: Web/packages/kernel/test/strings-conformance.test.ts
//
//  strings-conformance.test.ts — the localization kernel seam against the SHARED corpus
//  OpenSource/Conformance/strings/cases.json (architecture/localization.md). The Kotlin
//  twin runs the same file in :core (StringsConformanceTest); the Swift twin runs it on
//  the record lane. The seams are driven directly: state is the case's dot-keyed map,
//  the loader serves the case's bundle-table text, the device language is the case's.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";

import { DSXStrings } from "../src/strings.ts";

type Step = { state?: { [path: string]: unknown }; input: string; expect: string };
type Case = {
  name: string;
  device?: string;
  state?: { [path: string]: unknown };
  tables?: { [tag: string]: string };
  input?: string;
  expect?: string;
  steps?: Step[];
};

const corpus = JSON.parse(readFileSync(
  new URL("../../../../Conformance/strings/cases.json", import.meta.url),
  "utf8",
)) as { cases: Case[] };

test("the strings corpus is present and non-trivial", () => {
  assert.ok(corpus.cases.length >= 15, `expected the full case set, found ${corpus.cases.length}`);
});

for (const c of corpus.cases) {
  test(`strings corpus: ${c.name}`, () => {
    const state: { [path: string]: unknown } = { ...(c.state ?? {}) };
    const previous = { statePath: DSXStrings.statePath, loader: DSXStrings.loader, deviceLang: DSXStrings.deviceLang };
    DSXStrings.reset();
    DSXStrings.statePath = (path) => (path in state ? state[path] : null);
    DSXStrings.loader = (lang) => c.tables?.[lang] ?? null;
    DSXStrings.deviceLang = c.device ?? "en";
    try {
      const steps: Step[] = c.steps ?? [{ input: c.input!, expect: c.expect! }];
      for (const [index, step] of steps.entries()) {
        Object.assign(state, step.state ?? {});
        assert.equal(DSXStrings.localize(step.input), step.expect, `step ${index + 1}`);
      }
    } finally {
      DSXStrings.statePath = previous.statePath;
      DSXStrings.loader = previous.loader;
      DSXStrings.deviceLang = previous.deviceLang;
      DSXStrings.reset();
    }
  });
}
